Failed build on Windows.

Hi all,

I've had success with installing Simbody on a few environments, but I'm getting stuck on trying to compile on my local Windows machine. Any help would be greatly appreciated!

**Basic information**:
* [Up-to-date Simbody](https://github.com/simbody/simbody/commit/dda623d3e099b38ef9e551a6b647cb5e8dfbbb25)
* CMake 3.19.0-rc3
* Visual Studio 16 2019
* RelWithDebInfo

**CMake Output**:
```
CMake Deprecation Warning at CMakeLists.txt:19 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19042.
The C compiler identification is MSVC 19.28.29334.0
The CXX compiler identification is MSVC 19.28.29334.0
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
Found OpenGL: opengl32   
Configuring done
Generating done
```

**Snippet of Visual Studio Output** (focusing on the error):
```
Rebuild started...
1>------ Rebuild All started: Project: ZERO_CHECK, Configuration: RelWithDebInfo x64 ------
1>Checking Build System
2>------ Rebuild All started: Project: PlatformFiles_subdir, Configuration: RelWithDebInfo x64 ------
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libblas.dll -> C:\LIB\simbody\build\RelWithDebInfo\libblas.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libgcc_s_sjlj-1.dll -> C:\LIB\simbody\build\RelWithDebInfo\libgcc_s_sjlj-1.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libgfortran-3.dll -> C:\LIB\simbody\build\RelWithDebInfo\libgfortran-3.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\liblapack.dll -> C:\LIB\simbody\build\RelWithDebInfo\liblapack.dll
2>Copy C:\LIB\simbody\Platform\Windows\lib_x64\libquadmath-0.dll -> C:\LIB\simbody\build\RelWithDebInfo\libquadmath-0.dll
2>Building Custom Rule C:/LIB/simbody/Platform/CMakeLists.txt
3>------ Rebuild All started: Project: Code - Platform Files, Configuration: RelWithDebInfo x64 ------
3>Building Custom Rule C:/LIB/simbody/CMakeLists.txt
4>------ Rebuild All started: Project: Code - SimTKcommon, Configuration: RelWithDebInfo x64 ------
4>Building Custom Rule C:/LIB/simbody/SimTKcommon/sharedTarget/CMakeLists.txt
4>About.cpp
4>NiceTypeName.cpp

...
...
...

5>Geo_Point.cpp
5>Geo_Sphere.cpp
5>Geo_Triangle.cpp
5>Geodesic.cpp
5>OBBTree.cpp
5>OrientedBoundingBox.cpp
5>gcvspl.cpp
5>About.cpp
5>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\algorithm(7417,1): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
5>        with
5>        [
5>            _BidIt=SimTK::VectorIterator<SimTK::Real,SimTK::VectorBase<SimTK::Real>>
5>        ] (compiling source file C:\LIB\simbody\SimTKmath\Geometry\src\ContactGeometry_TriangleMesh.cpp)

...
...
...



31>TestScalar.cpp
29>TestPolygonalMesh.cpp
35>TestXml.cpp
16>TestArray.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\TestArray.exe
26>TestParallelExecutor.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\TestParallelExecutor.exe
11>RandomTest.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\RandomTest.exe
8>MatVecTest.vcxproj -> C:\LIB\simbody\build\RelWithDebInfo\MatVecTest.exe
5>Done building project "SimTKmath.vcxproj" -- FAILED.

```

Any help would be greatly appreciated. I can provide more information if needed.

If I'm reading this correctly, part of the problem appears in `ContactGeometry_TriangleMesh.cpp`:

```
5>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29333\include\algorithm(7417,1): error C2678: binary '*': no operator found which takes a left-hand operand of type 'const _BidIt' (or there is no acceptable conversion)
5>        with
5>        [
5>            _BidIt=SimTK::VectorIterator<SimTK::Real,SimTK::VectorBase<SimTK::Real>>
5>        ] (compiling source file C:\LIB\simbody\SimTKmath\Geometry\src\ContactGeometry_TriangleMesh.cpp)
5>C:\LIB\simbody\SimTKcommon\BigMatrix\include\SimTKcommon/internal/VectorIterator.h(76,10): message : could be 'ELT &SimTK::VectorIterator<ELT,SimTK::VectorBase<ELT>>::operator *(void)'
5>        with
5>        [
5>            ELT=SimTK::Real
5>        ] (compiling source file C:\LIB\simbody\SimTKmath\Geometry\src\ContactGeometry_TriangleMesh.cpp)
```

--- a/.github/workflows/continuous_integration.yml
+++ b/.github/workflows/continuous_integration.yml
@@ -30,7 +30,7 @@ jobs:
         run: |
           (New-Object System.Net.WebClient).DownloadFile("https://sourceforge.net/projects/myosin/files/doxygen-1.8.14.windows.x64.bin.zip/download", "doxygen.zip")
           7z x $env:GITHUB_WORKSPACE/doxygen.zip -odoxygen
-          echo "::add-path::$env:GITHUB_WORKSPACE\\doxygen"
+          echo "$env:GITHUB_WORKSPACE\\doxygen" >> $GITHUB_PATH
 
       - name: Configure
         run: |

--- a/.github/workflows/continuous_integration.yml
+++ b/.github/workflows/continuous_integration.yml
@@ -20,12 +20,16 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      # this is so deprecated ::add-path:: will still work (temporary fix)
+      - name: ACTIONS_ALLOW_UNSECURE_COMMANDS
+        id: ACTIONS_ALLOW_UNSECURE_COMMANDS
+        run: echo 'ACTIONS_ALLOW_UNSECURE_COMMANDS=true' >> $GITHUB_ENV
+
       - name: Install Doxygen
         # choco install doxygen.portable # <-- too unreliable.
         run: |
           (New-Object System.Net.WebClient).DownloadFile("https://sourceforge.net/projects/myosin/files/doxygen-1.8.14.windows.x64.bin.zip/download", "doxygen.zip")
           7z x $env:GITHUB_WORKSPACE/doxygen.zip -odoxygen
-          echo "ACTIONS_ALLOW_UNSECURE_COMMANDS=true" >> $GITHUB_ENV
           echo "::add-path::$env:GITHUB_WORKSPACE\\doxygen"
 
       - name: Configure

--- a/.github/workflows/continuous_integration.yml
+++ b/.github/workflows/continuous_integration.yml
@@ -25,6 +25,7 @@ jobs:
         run: |
           (New-Object System.Net.WebClient).DownloadFile("https://sourceforge.net/projects/myosin/files/doxygen-1.8.14.windows.x64.bin.zip/download", "doxygen.zip")
           7z x $env:GITHUB_WORKSPACE/doxygen.zip -odoxygen
+          echo "ACTIONS_ALLOW_UNSECURE_COMMANDS=true" >> $GITHUB_ENV
           echo "::add-path::$env:GITHUB_WORKSPACE\\doxygen"
 
       - name: Configure

--- a/SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h
+++ b/SimTKcommon/BigMatrix/include/SimTKcommon/internal/VectorIterator.h
@@ -73,14 +73,14 @@ public:
     same vector as the source iterator. **/
     VectorIterator& operator=(const VectorIterator& source) = default;
 
-    ELT& operator*() {
+    /** Dereference the iterator to access the element it points to. Note
+    that although the iterator may be const, it can still point to a 
+    mutable element. **/
+    ELT& operator*() const {
         assert (index >= 0 && index < vectorp->size());
         return (*vectorp)[(int)index];
     }
-    ELT& operator[](ptrdiff_t i) {
-        assert (i >= 0 && i < vectorp->size());
-        return (*vectorp)[(int)i];
-    }
+
     VectorIterator& operator++() {
         assert (index < vectorp->size());
         ++index;

--- a/SimTKcommon/Polynomial/src/cpoly.cpp
+++ b/SimTKcommon/Polynomial/src/cpoly.cpp
@@ -602,7 +602,7 @@ T CPoly<T>::scale( const int nn, const T pt[], const T eta, const T infin, const
       if( infin / sc > max ) sc = 1;
       }
    l = (int)( log( sc ) / log(base ) + 0.5 );
-   fn_val = pow( base , l );
+   fn_val = T(std::pow( base , l )); // extraneous cast required by VS 16.8.2
    return fn_val;
    }
 

--- a/SimTKcommon/Polynomial/src/rpoly.cpp
+++ b/SimTKcommon/Polynomial/src/rpoly.cpp
@@ -143,7 +143,7 @@ _40:
             sc = smalno;
     }
     l = (int)(log(sc)/log(base) + 0.5);
-    factor = pow(base*(T) 1.0,l);
+    factor = T(std::pow( base, l )); // extraneous cast required by VS 16.8.2
     if (factor != 1.0) {
         for (i=0;i<=n;i++) 
             p[i] = factor*p[i];     /* Scale polynomial. */

